pyro/infer/mcmc/hmc.py

```diff
@@ -398,6 +415,7 @@ def sample(self, trace):
         if rand < accept_prob:
             self._accept_cnt += 1
             z = z_new
+            self._cache(potential_energy_new, z_grads_new)
```

Comment:
We should cache pe and its grad here.

Reply:
Good catch! It is not doing any caching across sample steps since we recompute if the gradients / pe computation is None. That's why the tests weren't failing, but we weren't making use of this optimization for HMC.